Add CO2SOL and H2SOL

totto82 opened the pull request.

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

akva2 commented:
i'm have no idea why but jenkins doesn't see your requests here for some reason. i'll try.

akva2 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

flammmes commented:
How will CO2SOL be utilized? Is convective dissolution with brine calculated as in CO2STORE from the same correlations? If this is the case, will CO2SOL be able to completely replace CO2STORE especially in cases where we simulate CO2-EOR and subsequently CCS?  

Sorry if my question is out of place and if the answer is obvious by checking the commit, but I am still trying to figure out how the source code of OPM works.

totto82 commented:
> How will CO2SOL be utilized? Is convective dissolution with brine calculated as in CO2STORE from the same correlations? If this is the case, will CO2SOL be able to completely replace CO2STORE especially in cases where we simulate CO2-EOR and subsequently CCS?

The idea is that CO2SOL should be used in combination with the SOLVENT model to model CO2 injection in a hydrocarbon field. CO2STORE is for CO2 injection in an aquifer. CO2SOL is based on the same co2brine fluidsystem as CO2STORE. So CO2 dissolution in brine is modeled the same way. If you want to model CO2EOR using CO2SOL you first setup your basis blackoil model for the hydrocarbon gas/oil + water system using table input for the PVT. Then you add CO2SOL and SOLVENT to add CO2 as an additional gas component. Only the solvent (CO2) is allowed to dissolve into water/brine. The mixing of CO2 and the hydrocarbons (both the oil and gas phase) is modeled using the MISCIBILITY option in the solvent model.

flammmes commented:
> > How will CO2SOL be utilized? Is convective dissolution with brine calculated as in CO2STORE from the same correlations? If this is the case, will CO2SOL be able to completely replace CO2STORE especially in cases where we simulate CO2-EOR and subsequently CCS?
> 
> The idea is that CO2SOL should be used in combination with the SOLVENT model to model CO2 injection in a hydrocarbon field. CO2STORE is for CO2 injection in an aquifer. CO2SOL is based on the same co2brine fluidsystem as CO2STORE. So CO2 dissolution in brine is modeled the same way. If you want to model CO2EOR using CO2SOL you first setup your basis blackoil model for the hydrocarbon gas/oil + water system using table input for the PVT. Then you add CO2SOL and SOLVENT to add CO2 as an additional gas component. Only the solvent (CO2) is allowed to dissolve into water/brine. The mixing of CO2 and the hydrocarbons (both the oil and gas phase) is modeled using the MISCIBILITY option in the solvent model.

Thank you very much for your reply. I went on to pull this and the other two related requests in opm-models and opm-simulators and build them.
When I tried to build opm-simulators with #4991 I got the following error.

[  1%] Building CXX object CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc: In member function ‘void Opm::EclGenericOutputBlackoilModule<FluidSystem, Scalar>::setRestart(const Opm::data::Solution&, unsigned int, unsigned int)’:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: error: no matching function for call to ‘Opm::data::Solution::data(const char [7]) const’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note: candidate: ‘template<class T> std::vector<Scalar>& Opm::data::Solution::data(const string&)’
   51 |         std::vector<T>& data(const std::string& );
      |                         ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note: candidate: ‘template<class T> const std::vector<Scalar>& Opm::data::Solution::data(const string&) const’
   54 |         const std::vector<T>& data(const std::string& ) const;
      |                               ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
make[2]: *** [CMakeFiles/opmsimulators.dir/build.make:104: CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:338: CMakeFiles/opmsimulators.dir/all] Error 2
make: *** [Makefile:146: all] Error 2 

So I changed 658 to
rswSol_[elemIdx] = sol.data<Scalar>("RSWSOL")[globalDofIndex]; 
and now it builds fine.

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

svenn-t commented on src/opm/material/fluidsystems/BlackOilFluidSystem.cpp:
```diff
@@ -84,9 +84,11 @@ initFromState(const EclipseState& eclState, const Schedule& schedule)
         if (eclState.runspec().co2Storage() || eclState.runspec().h2Storage())
             setEnableDissolvedGasInWater(eclState.getSimulationConfig().hasDISGASW());
-        else
+        else if (eclState.runspec().co2Sol() || eclState.runspec().h2Sol()) {
+
+        } else
```
There is something missing here.

totto82 replied:
I have added a message to the user

totto82 commented:
I have updated the PR with some more messages. And more fundamentally I have switched to using the water properties from the brine module to make it consistent with the properties used by the solvent model.

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please

totto82 commented:
jenkins build this opm-models=851 opm-simulators=4991 please
